de.metas.business/src/main/java/de/metas/customs/CustomsInvoiceService.java

```diff
@@ -0,0 +1,258 @@
+		{
+			documentBL.processEx(customsInvoiceRecord, IDocument.ACTION_Complete, IDocument.STATUS_Completed);
+		}
+
+		return customsInvoice;
+	}
+
+	private CustomsInvoice createCustomsInvoice(final BPartnerLocationId bpartnerAndLocationId,
+			final UserId userId,
+			final DocTypeId docTypeId,
+			final String documentNo,
+			final CurrencyId currencyId,
+			final LocalDate invoiceDate,
+			final Map<ProductId, List<InOutAndLineId>> linesToExportMap)
+	{
+
+		final ImmutableList<CustomsInvoiceLine> customsInvoiceLines = linesToExportMap.keySet()
+				.stream()
+				.map(productId -> createCustomsInvoiceLine(productId, linesToExportMap.get(productId), currencyId))
+				.collect(ImmutableList.toImmutableList());
+
+		final CustomsInvoice customsInvoice = CustomsInvoice.builder()
+				.bpartnerAndLocationId(bpartnerAndLocationId)
+				.userId(userId)
+				.invoiceDate(invoiceDate)
+				.orgId(Env.getOrgId())
+				.currencyId(currencyId)
+				.docStatus(IDocument.STATUS_Drafted)
+				.docAction(IDocument.ACTION_Complete)
+				.docTypeId(docTypeId)
+				.documentNo(documentNo)
+				.lines(customsInvoiceLines)
+				.build();
+
+		customsInvoice.updateLineNos();
+
+		return customsInvoice;
+	}
+
+	private CustomsInvoiceLine createCustomsInvoiceLine(final ProductId productId,
+			final List<InOutAndLineId> shipmentLinesForProducts,
+			final CurrencyId currencyId)
+	{
+		final IProductDAO productDAO = Services.get(IProductDAO.class);
+
+		final I_M_Product product = productDAO.getById(productId);;
+
+		Quantity qty = Quantity.of(BigDecimal.ZERO, product.getC_UOM());
+		Money lineNetAmt = Money.of(BigDecimal.ZERO, currencyId);
+
+		final UomId uomId = UomId.ofRepoId(product.getC_UOM_ID());
+
+		for (InOutAndLineId inoutAndLineId : shipmentLinesForProducts)
+		{
+			final Quantity inoutLineQtyCoverted = getInOutLineQtyConverted(inoutAndLineId, uomId);
+
+			qty = qty.add(inoutLineQtyCoverted.getAsBigDecimal());
+
+			final Money inoutLinePriceConverted = getInOutLinePriceConverted(inoutAndLineId, currencyId);
+
+			final Money shipmentLineNetAmt = inoutLinePriceConverted.multiply(inoutLineQtyCoverted.getAsBigDecimal());
+
+			lineNetAmt = lineNetAmt.add(shipmentLineNetAmt);
+
+		}
+
+		final CustomsInvoiceLine customsInvoiceLine = CustomsInvoiceLine.builder()
+				.productId(productId)
+				.lineNetAmt(lineNetAmt)
+				.quantity(qty)
+				.orgId(Env.getOrgId())
+				.uomId(uomId)
+				.build();
+
+		return customsInvoiceLine;
+	}
+
+	private Money getInOutLinePriceConverted(final InOutAndLineId inoutAndLineId, final CurrencyId currencyId)
+	{
+		final IInOutDAO inoutDAO = Services.get(IInOutDAO.class);
+		final ICurrencyBL currencyBL = Services.get(ICurrencyBL.class);
+
+		final InOutLineId inOutLineId = inoutAndLineId.getInOutLineId();
+
+		final I_M_InOutLine inoutLineRecord = inoutDAO.getLineById(inOutLineId);
+
+		final OrderLineId orderLineId = OrderLineId.ofRepoIdOrNull(inoutLineRecord.getC_OrderLine_ID());
+
+		if (Check.isEmpty(orderLineId))
```

Comment:
Did you check with mark if this is OK? Otherwise, you might need to fall back to the invoice candidate, or maybe using the pricing engine..
After all, they might want to export something that is just on the shipment line and has no order line. btw, what about packaging inout lines? They don't have an order lines

EDIT: OK, so it seems as if packaging lines are both going to end up in a customs invoice. still, pls clarify with mark..

Reply:
He said it's ok for the first round. He will ask the customer tomorrow